server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/bitcoin/exchanges/bitbuy/BitbuyExchange.java

```diff
@@ -0,0 +1,329 @@
+import si.mazi.rescu.HttpStatusIOException;
+
+import java.math.BigDecimal;
+import java.security.GeneralSecurityException;
+import java.util.Set;
+import java.util.concurrent.Callable;
+
+public class BitbuyExchange implements IExchangeAdvanced, IRateSourceAdvanced {
+    private static final Logger log = LoggerFactory.getLogger("batm.master.exchange.BitbuyExchange");
+
+    private static final Set<String> fiatCurrencies = ImmutableSet.of(
+        FiatCurrency.CAD.getCode(),
+        CryptoCurrency.DAI.getCode()); // stable coin as fiat
+
+    private static final Set<String> cryptoCurrencies = ImmutableSet.of(
+        CryptoCurrency.BCH.getCode(),
+        CryptoCurrency.BTC.getCode(),
+        CryptoCurrency.DAI.getCode(),
+        CryptoCurrency.ETH.getCode(),
+        CryptoCurrency.LTC.getCode(),
+        CryptoCurrency.XRP.getCode());
+
+    // Supported markets (2021-10-06; see IBitbuyAPI.getMarkets):
+    // [DAI-CAD, ETH-DAI, BTC-AAVE, ETH-AAVE, ETH-CAD, BTC-LINK, ETH-LINK, BCH-CAD, AAVE-CAD, BCH-BTC, BTC-DAI, LINK-CAD, BTC-CAD, XLM-CAD, LTC-CAD, XRP-CAD, EOS-BTC, XLM-BTC, XRP-BTC, ETH-BTC, LTC-BTC, EOS-CAD]
+
+    private final String preferredFiatCurrency;
+    private final IBitbuyAPI api;
+
+    public BitbuyExchange(String apiKey, String apiSecret, String preferredFiatCurrency) throws GeneralSecurityException {
+        this.api = IBitbuyAPI.create(apiKey, apiSecret);
+        this.preferredFiatCurrency = preferredFiatCurrency;
+    }
+
+    private String getMarketSymbol(String cryptoCurrency, String fiatCurrency) {
+        return cryptoCurrency + "-" + fiatCurrency;
+    }
+
+    private boolean isCryptoCurrencySupported(String currency) {
+        if (!getCryptoCurrencies().contains(currency)) {
+            log.debug("doesn't support cryptocurrency '{}'", currency);
+            return false;
+        }
+        return true;
+    }
+
+    private boolean isFiatCurrencySupported(String currency) {
+        if (!getFiatCurrencies().contains(currency)) {
+            log.debug("doesn't support fiat currency '{}'", currency);
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public BigDecimal getCryptoBalance(String cryptoCurrency) {
+        if (!isCryptoCurrencySupported(cryptoCurrency)) {
+            return BigDecimal.ZERO;
+        }
+        return getBalance(cryptoCurrency);
+    }
+
+    @Override
+    public BigDecimal getFiatBalance(String fiatCurrency) {
+        if (!isFiatCurrencySupported(fiatCurrency)) {
+            return BigDecimal.ZERO;
+        }
+        return getBalance(fiatCurrency);
+    }
+
+    private BigDecimal getBalance(String currency) {
+        return call(currency + " balance", () ->
+            api.getWallets().stream()
+                .filter(w -> currency.equals(w.symbol))
+                .findAny()
+                .map(w -> w.availableBalance)
+                .orElseThrow(() -> new Exception(currency + " wallet not found"))
```

Comment:
Does it have to be `Exception`, what about just `RuntimeException`? 

Btw you could also rename `w`  to `wallet` but that's not so important